Multiple drivers

jedwards4b opened the pull request:
Implementation of multiple MCT drivers as an option for multi-instance simulations. If multi-instance is enabled, N drivers are run, each with one instance.

Also (changes not directly related to multi-driver):

- Changed interface of `check_lockedfiles` (check_lockedfiles.py) to take a case instead of a caseroot.
- Use `case.get_env` instead of `EnvBuild` in check_lockedfiles.py
- Changed `check_case` (case_submit.py) to not take a caseroot input.
- Cleaned up memleak testing in `_check_for_memleak` (system_tests_common.py)
- Fixed bad format in `build_xcpl_nml` (buildnml.py)

Test suite: scripts_regression_tests.py
Test baseline: NA
Test namelist changes: NA
Test status: bit for bit
Fixes: #1704 
Fixes: #1714 

User interface changes?: new --multi-driver option to create_newcase and _C# modifier to tests

Update gh-pages html (Y/N)?: Y

Code review: @gold2718

gold2718 commented on scripts/lib/CIME/SystemTests/pre.py:
```diff
@@ -85,26 +85,40 @@ def run_phase(self): # pylint: disable=arguments-differ
+                else:
+                    ninst = 1
+            else:
+                ninst = self._case.get_value("NINST_{}".format(comp.upper()))
```
Should there be a blank line here (end of `if`)?

gold2718 commented on scripts/lib/CIME/SystemTests/pre.py:
```diff
@@ -85,26 +85,40 @@ def run_phase(self): # pylint: disable=arguments-differ
+                ok = cprnc(comp, rfile1, rfile2, self._case, rundir2)[0]
+                logger.warning("CPRNC result for {}: {}".format(os.path.basename(rfile1), "PASS" if (ok == should_match) else "FAIL"))
+                compare_ok = compare_ok and (should_match == ok)
 
```
I'm not sure what's wrong yet but both `./create_test PRE.f09_g17.ADESP.hobart_nag.drv-som` and `./create_test PRE_N3.f09_g17.ADESP.hobart_nag.drv-som` fail so I'm going to pause the review here until I get an idea of what's wrong.

jedwards4b replied:
The issue is the ocn mask as was discussed earlier in the week wrt the master.  I don't think that this failure is unique to this branch is it?

jedwards4b replied:
The correct tests are  PRE.f09_f09.ADESP.hobart_nag PRE.f09_f09.ADESP_TEST.hobart_nag and they both pass.

gold2718 replied:
Right. Also, it turns out the forcing file pointed to by drv-som is bad but that change was not in the commit message so I got confused. I'm still looking into things. Currently:
```
PRE_N3.f19_f19.ADESP.hobart_nag
```
passes but:
```
PRE_C3.f19_f19.ADESP.hobart_nag
```
is failing. Still have lots of other tests to do as well.

jedwards4b replied:
Working on this one now.

jedwards4b replied:
This test is working now.

jedwards4b commented on config/config_tests.xml:
```diff
@@ -500,6 +500,17 @@ NODEFAIL          Tests restart upon detected node failure. Generates fake failu
+    <HIST_N>$STOP_N</HIST_N>
+    <MULTI_DRIVER>TRUE</MULTI_DRIVER>
+  </test>
+
```
The comment on line 128 has nothing to do with this PR.  If it isn't correct please open an issue.   I think that adding more tests to the system can come after we accept this functionality PR and should also be a separate issue.

jedwards4b commented on doc/source/users_guide/multi-instance.rst:
```diff
@@ -47,16 +45,21 @@ To run two instances of CAM, CLM, CICE, RTM and DOCN, invoke the following comma
    > ./case.setup
 
-A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instances) is generated when **case.setup** is called. 
+A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instances) is generated when **case.setup** is called.
```
that's not true  - it is correct as written.

jedwards4b commented on doc/source/users_guide/multi-instance.rst:
```diff
@@ -47,16 +45,21 @@ To run two instances of CAM, CLM, CICE, RTM and DOCN, invoke the following comma
-A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instances) is generated when **case.setup** is called. 
+A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instances) is generated when **case.setup** is called.
 When calling **case.setup** with the **env_mach_pes.xml** file specifically, these files are created in **$CASEROOT**:
 ::
```
I don't understand why you are commenting on code and comments that I haven't changed.   I think you are correct but are they really issues for this PR or should they be noted separately?

jedwards4b commented:
Rebased to master, @gold2718 awaiting your feedback.

gold2718 commented on scripts/lib/CIME/check_lockedfiles.py:
```diff
@@ -122,15 +123,13 @@ def check_lockedfiles(caseroot=None):
                     else:
-                        f1obj.set_value("BUILD_STATUS", 1)
-                        f1obj.write()
+                        case.set_value("BUILD_STATUS", 1)
```
Are the changes in the module even related to multi_driver? If not, please open a new issue and submit a PR with these changes.

jedwards4b replied:
I don't remember and think that its a total waste of time to try to figure that out.

gold2718 replied:
Well I do not understand why you have a double standard where it is okay to fix incidental _code_ issues found when working on a new feature but not okay to fix incidental _comment_ or _documentation_ issues found in the same process.
How about:

1. Fix the documentation/comment issues I pointed out
2. Include information about the incidental fixes in the PR text so that there is some record in the log that this was done.

and I will drop my request for proper separation between new features and bug fixes.

gold2718 commented on scripts/lib/CIME/buildnml.py:
```diff
@@ -78,7 +78,7 @@ def build_xcpl_nml(case, caseroot, compname):
             filename = os.path.join(rundir, "{}_in".format(compname))
         else:
-            filename = os.path.join(rundir, "{}_in_{:4.4d}".format(compname, i))
+            filename = os.path.join(rundir, "{}_in_{:04d}".format(compname, i))
```
This seems like an unrelated bugfix which should be in its own PR.

gold2718 commented on scripts/lib/CIME/case_run.py:
```diff
@@ -244,7 +257,7 @@ def case_run(case, skip_pnl=False):
            "As a result, short-term archiving will not be called automatically."
            "Please submit your run using the submit script like so:"
-           " ./case.submit")
+           " ./case.submit Time: {}".format(get_timestamp()))
```
Is this change related to multi_driver? If not, please open an issue and submit it as a PR.

gold2718 commented on scripts/lib/CIME/get_timing.py:
```diff
@@ -150,11 +162,16 @@ def getTiming(self):
 
-        binfilename = os.path.join(rundir, "timing", "model_timing_stats")
+        if inst > 0:
+            inst_label = '_%04d' % inst
```
Shouldn't this use the `'_{:04d}'.format(inst)` syntax?

gold2718 commented on scripts/lib/CIME/get_timing.py:
```diff
@@ -150,11 +162,16 @@ def getTiming(self):
+        else:
+            inst_label = ''
+
+        binfilename = os.path.join(rundir, "timing", "model_timing%s_stats" % inst_label)
```
Shouldn't this use the `"model_timing{}_stats".format(inst_label)` syntax?

gold2718 commented on src/build_scripts/buildlib.csm_share:
```diff
@@ -19,14 +19,27 @@ my $USE_ESMF_LIB	= `./xmlquery USE_ESMF_LIB	--value`;
+
+if ($multi_driver eq "FALSE") {
+    $NINST_ATM		= `./xmlquery NINST_ATM		--value`;
+    $NINST_ICE		        = `./xmlquery NINST_ICE		--value`;
```
Tab character sneaked through? EXTERMINATE!

gold2718 commented on src/components/data_comps/desp/desp_comp_mod.F90:
```diff
@@ -219,8 +219,13 @@ subroutine desp_comp_init(EClock, espid, mpicom_in, phase, read_restart,    &
+      call mpi_comm_rank(global_comm, global_mype, ierr)
+      call mpi_comm_size(global_comm, global_numpes, ierr)
+
+      write(logunit,*)'DESP: I am global rank ',global_mype,' of ',global_numpes
```
Is this write statement still necessary / useful?

jedwards4b replied:
Yes I think that it is and this code is only used in tests.

gold2718 commented on src/components/xcpl_comps/xshare/dead_mod.F90:
```diff
@@ -216,7 +216,7 @@ subroutine dead_setNewGrid(decomp_type,nxg,nyg,totpe,mype,lsize,gbuf,seg_len,npr
       enddo
 
-      write(logunit,*) 'dead_setNewGrid decomp seg ',mype,lsize,nx
+!      write(logunit,*) 'dead_setNewGrid decomp seg ',mype,lsize,nx
```
I agree, let's remove this.

jedwards4b replied:
done

gold2718 commented on src/drivers/mct/cime_config/config_component.xml:
```diff
@@ -1949,6 +1949,18 @@
+    <valid_values>TRUE,FALSE</valid_values>
+    <group>mach_pes</group>
+    <file>env_mach_pes.xml</file>
+    <desc>MULTI_DRIVER mode provides a separate driver (and coupler) component for each
```
I would rephrase this line as:
<desc>MULTI_DRIVER mode provides a separate driver (and coupler component) for each

gold2718 commented on src/drivers/mct/main/cime_comp_mod.F90:
```diff
@@ -843,7 +855,12 @@ subroutine cime_pre_init2()
    !----------------------------------------------------------
 
-   call seq_infodata_init(infodata,nlfilename, GLOID, pioid)
+   if (len(cpl_inst_tag) > 0) then
```
Shouldn't this be `len_trim` instead of `len`?

gold2718 commented on src/drivers/mct/main/cime_comp_mod.F90:
```diff
@@ -3573,7 +3587,7 @@ subroutine cime_run()
               atm, lnd, ice, ocn, rof, glc, wav, esp,                 &
               fractions_ax, fractions_lx, fractions_ix, fractions_ox, &
-              fractions_rx, fractions_gx, fractions_wx)
+              fractions_rx, fractions_gx, fractions_wx, tag=trim(cpl_inst_tag))
```
Why specify the name of a non-optional input? Seems inconsistent.

jedwards4b replied:
fixed

jedwards4b commented on doc/source/users_guide/multi-instance.rst:
```diff
@@ -47,16 +45,21 @@ To run two instances of CAM, CLM, CICE, RTM and DOCN, invoke the following comma
 ::
 
    > ./case.setup
 
```
works for me

jedwards4b commented on doc/source/users_guide/multi-instance.rst:
```diff
@@ -47,16 +45,21 @@ To run two instances of CAM, CLM, CICE, RTM and DOCN, invoke the following comma
    > ./case.setup
 
-A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instances) is generated when **case.setup** is called. 
+A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instances) is generated when **case.setup** is called.
```
Whether or not multi_driver does not change the result here.   I still maintain that this is correct as written.

gold2718 replied:
I tried it and got different files depending on whether multi_driver was true or not.
Also, it cannot be correct as written as it refers to a single file but multiple files are created.

jedwards4b replied:
I just tried this again and got exactly the same result - perhaps if you are getting a different result you should document exactly what that difference is.

jedwards4b replied:
```

jedwards4b replied:
```
./xmlchange MULTI_DRIVER=TRUE
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) rm user_nl*
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) ./case.setup --reset
Successfully cleaned batch script case.run
Setting resource.RLIMIT_STACK to -1 from (307200000, -1)
Creating batch script case.run
Writing case.run script from input template /glade/p/work/jedwards/sandboxes/cesm2_0_alpha/cime.jet/config/cesm/machines/template.case.run
Writing case.st_archive script from input template /glade/p/work/jedwards/sandboxes/cesm2_0_alpha/cime.jet/config/cesm/machines/template.st_archive
Locking file env_mach_pes.xml
Creating user_nl_xxx files for components and cpl
If an old case build already exists, might want to run 'case.build --clean' before building
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) ls -ltr user*
-rw-r--r-- 1 jedwards ncar  848 Sep  1 11:28 user_nl_cpl
-rw-r--r-- 1 jedwards ncar  115 Sep  1 11:28 user_nl_cam_0001
-rw-r--r-- 1 jedwards ncar  115 Sep  1 11:28 user_nl_cam_0002
-rw-r--r-- 1 jedwards ncar 1489 Sep  1 11:28 user_nl_clm_0001
-rw-r--r-- 1 jedwards ncar 1489 Sep  1 11:28 user_nl_clm_0002
-rw-r--r-- 1 jedwards ncar  367 Sep  1 11:28 user_nl_cice_0001
-rw-r--r-- 1 jedwards ncar  367 Sep  1 11:28 user_nl_cice_0002
-rw-r--r-- 1 jedwards ncar  932 Sep  1 11:28 user_nl_docn_0001
-rw-r--r-- 1 jedwards ncar  932 Sep  1 11:28 user_nl_docn_0002
-rw-r--r-- 1 jedwards ncar  595 Sep  1 11:28 user_nl_mosart_0001
-rw-r--r-- 1 jedwards ncar  595 Sep  1 11:28 user_nl_mosart_0002
-rw-r--r-- 1 jedwards ncar  781 Sep  1 11:28 user_nl_cism_0001
-rw-r--r-- 1 jedwards ncar  781 Sep  1 11:28 user_nl_cism_0002
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) rm user_nl*
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) ./xmlchange MULTI_DRIVER=False
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) ./case.setup --reset
Successfully cleaned batch script case.run
Setting resource.RLIMIT_STACK to -1 from (307200000, -1)
Creating batch script case.run
Writing case.run script from input template /glade/p/work/jedwards/sandboxes/cesm2_0_alpha/cime.jet/config/cesm/machines/template.case.run
Writing case.st_archive script from input template /glade/p/work/jedwards/sandboxes/cesm2_0_alpha/cime.jet/config/cesm/machines/template.st_archive
Locking file env_mach_pes.xml
Creating user_nl_xxx files for components and cpl
If an old case build already exists, might want to run 'case.build --clean' before building
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) ls -ltr user*
-rw-r--r-- 1 jedwards ncar  848 Sep  1 11:29 user_nl_cpl
-rw-r--r-- 1 jedwards ncar  115 Sep  1 11:29 user_nl_cam_0001
-rw-r--r-- 1 jedwards ncar  115 Sep  1 11:29 user_nl_cam_0002
-rw-r--r-- 1 jedwards ncar 1489 Sep  1 11:29 user_nl_clm_0001
-rw-r--r-- 1 jedwards ncar 1489 Sep  1 11:29 user_nl_clm_0002
-rw-r--r-- 1 jedwards ncar  367 Sep  1 11:29 user_nl_cice_0001
-rw-r--r-- 1 jedwards ncar  367 Sep  1 11:29 user_nl_cice_0002
-rw-r--r-- 1 jedwards ncar  932 Sep  1 11:29 user_nl_docn_0001
-rw-r--r-- 1 jedwards ncar  932 Sep  1 11:29 user_nl_docn_0002
-rw-r--r-- 1 jedwards ncar  595 Sep  1 11:29 user_nl_mosart_0001
-rw-r--r-- 1 jedwards ncar  595 Sep  1 11:29 user_nl_mosart_0002
-rw-r--r-- 1 jedwards ncar  781 Sep  1 11:29 user_nl_cism_0001
-rw-r--r-- 1 jedwards ncar  781 Sep  1 11:29 user_nl_cism_0002
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
```

jedwards4b commented on doc/source/users_guide/multi-instance.rst:
```diff
@@ -79,17 +82,21 @@ Also, **case.setup** creates the following ``*_in_*`` files and ``*txt*`` files
    lnd_in_0001, lnd_in_0002
    rof_in_0001, rof_in_0002
 
```
Removed this erroneous statement.

jedwards4b commented on scripts/Tools/check_case:
```diff
@@ -50,9 +50,8 @@ def _main_func(description):
-    check_lockedfiles()
-
     with Case(read_only=False) as case:
+        check_lockedfiles(case)
```
I cannot sort out at this point what is related to multi-driver and what is not.   This are problems that I found and fixed in the course of developing this branch.

jedwards4b commented on scripts/lib/CIME/SystemTests/system_tests_common.py:
```diff
@@ -349,6 +352,24 @@ def compare_env_run(self, expected=None):
+            for log in cpllogs:
+                if log in lastcpllogs:
+                    continue
+                if log.endswith(suffix):
```
space added

jedwards4b commented on scripts/lib/CIME/XML/env_mach_pes.py:
```diff
@@ -20,6 +20,13 @@ def __init__(self, case_root=None, infile="env_mach_pes.xml", components=None):
+            for comp in self._components:
+                value = max(value, self.get_value("NINST_{}".format(comp)))
+            return value
+
```
done

jedwards4b commented:
I have made most of the requested changes.

rljacob commented:
How about we call this "Ensemble drivers"?   "multiple drivers" to me is the work Mariana is doing for the NUOPC driver.

jedwards4b commented:
Mutiple drivers is meant to distinguish from the multiple components mode.  I think that ensemble drivers is even more confusing.

jedwards4b commented:
@gold2718 I'm having trouble following the flow of this conversation, if you still fell that I have not addressed issues please be explicit about exactly what those are.

gold2718 commented:
```
: ./create_newcase --case noMulti --compset F2000_DEV --res ne30_ne30_mg17
: cd noMulti
: ./xmlchange NINST_ATM=2,NINST_ROF=2,NINST_OCN=2,NINST_ICE=2
: ./xmlchange MULTI_DRIVER=FALSE
: ./case.setup
: ./preview_namelists
: ./xmlquery NINST
        NINST: ['ATM:2', 'LND:1', 'ICE:2', 'OCN:2', 'ROF:2', 'GLC:1', 'WAV:1', 'ESP:1']
: ls Fmulti/user_nl_*
Fmulti/user_nl_cam        Fmulti/user_nl_cice_0002  Fmulti/user_nl_docn_0001
Fmulti/user_nl_cam_0001   Fmulti/user_nl_cism       Fmulti/user_nl_docn_0002
Fmulti/user_nl_cam_0002   Fmulti/user_nl_clm        Fmulti/user_nl_mosart_0001
Fmulti/user_nl_cice_0001  Fmulti/user_nl_cpl        Fmulti/user_nl_mosart_0002

```
Note that there are no `_NNNN` files for CISM, CLM, or CPL.
Now, do the exact same thing with multi_driver
```
: ./create_newcase --case Multi --compset F2000_DEV --res ne30_ne30_mg17
: cd Multi
: ./xmlchange NINST_ATM=2,NINST_ROF=2,NINST_OCN=2,NINST_ICE=2
: ./xmlchange MULTI_DRIVER=TRUE
: ./case.setup
: ./preview_namelists
: ./xmlquery NINST
        NINST: ['ATM:2', 'LND:1', 'ICE:2', 'OCN:2', 'ROF:2', 'GLC:1', 'WAV:1', 'ESP:1']
: ls Fmulti/user_nl_*
user_nl_cam        user_nl_cice_0002  user_nl_clm_0001   user_nl_docn_0002
user_nl_cam_0001   user_nl_cism_0001  user_nl_clm_0002   user_nl_mosart_0001
user_nl_cam_0002   user_nl_cism_0002  user_nl_cpl        user_nl_mosart_0002
user_nl_cice_0001  user_nl_clm        user_nl_docn_0001
```
So either the documentation is wrong or the `NINST` report is wrong or `xmlchange` is not working properly.

jedwards4b commented:
I already updated the documentation in this PR to explicitly call out the additional changes to NINST:
```
   > ./xmlchange NINST_GLC=2                                                    
   > ./xmlchange NINST_LND=2            
```
so what are you looking at?

gold2718 commented:
I was not trying to follow the directions, I executed a series of commands and noted that the user_nl_xxx_NNNN files is not as described in multi-instance.rst.
I also wondered if the output is correct or if my command sequence should have generated an error.
Can you please respond to my post above?

gold2718 commented:
Another item which is not completed is that there is still no documentation block for the MCC test type in config_tests.xml.

jedwards4b commented:
So the document provides directions that you didn't follow and now you don't understand why you got a different result?   I've updated config_tests.xml

gold2718 commented:
Yes, it is a new technique called 'testing'. Either the commands I entered are correct and the documentation (stating which user_nl files are created) is wrong, or the commands I entered are illegal and the code did not respond appropriately.
I tend to think I entered an illegal set of commands and the software should have either auto-corrected, warned or generated an error but I'm open to other interpretations.

jedwards4b commented on doc/source/users_guide/multi-instance.rst:
```diff
@@ -1,95 +1,109 @@
 
-#. Multiple instances generally should un concurrently, which is the default setting in **env_mach_pes.xml**. 
-   The serial setting is only for EXPERT USERS in upcoming development code implementations.
+#. In multi-driver mode you will always get 1 instance of each component for each driver/coupler, if you change a case using xmlchange MULTI_COUPLER=TRUE you will get a number of driver/couplers equal to the maximum NINST value over all components.
```
@gold2718 this note was added to the documentation to clarify use of multi-driver mode.   In my opinion there is no error in what you did to create a case, there is however an expectation that the user understand how multi-driver differs from multi-instance and thus would expect a different list of user_nl_XXXX files.

gold2718 replied:
This seems confusion prone as `xmlquery NINST` does not show how many instances will actually run. @mvertens, @billsacks, any objections to this situation?

billsacks replied:
@gold2718 Sorry, I haven't been following this whole thread. I'd be happy to weigh in if you'd be able to give a brief summary of the situation that you want input on.

gold2718 replied:
Look back about 6 posts to yesterday where I show two sets of commands (beginning with `./create_newcase`) and show that when `MULTI_DRIVER=TRUE`, the number of instances shown by `./xmlquery NINST` is not necessarily the number of instances which will be run.

billsacks replied:
Sorry, I took some time to try to understand the current behavior and expected behavior, but I'm still a bit lost. Could we have a brief in-person chat about this on Tuesday?

gold2718 replied:
I don't want to take up a lot of your time with this but I will try a brief summary.
With a single driver, you can have N instances of some components but only one instance of others. For instance, you can have one ocean averaging the results of 64 atmospheres.

With multi-driver, each driver drives a separate instance of every non-stub component so if there are N drivers, there are N instances of every non-stub component.

The issue is that `xmlchange` allows you to set instance counts which are not viable with multi-driver but rather than display an error, the multi-driver code just ignores the situation and creates max(NINST_xxx) instances of every component. This behavior is mentioned at the end of the multi-instance documentation yet `xmlquery` still shows the values set by the user which is _not_ what is actually being run.

My question is whether this situation is okay or should be fixed. My concern is that users may read or hear that multi-driver speeds up the code so they turn it on. They can even check that they are running one ocean against 64 atmospheres (with `xmlquery`) but will in fact run 64 ocean instances with no averaging. As far as I can tell, there is no way to tell this is going to happen without running the case (or very carefully RTFM and we know how likely that is).

gold2718 replied:
If we decide that there should be a fix, I propose the following actions:

1. When setting `MULTI_DRIVER` from `FALSE` to `TRUE`, check to make sure all `NINST_xxx` have the same value and generate an error if not.
2. If `MULTI_DRIVER == TRUE`, then `case.setup` will check that all `NINST_xxx` have the same value and generate an error if not.

billsacks replied:
@gold2718 thanks for your explanation. Now I understand the issue. Yes, I agree with you that this basic sanity checking should be done, rather than silently running something different from what was requested and/or listed values would suggest.

gold2718 commented:
Aside from the `xmlquery NINST` possible issue, the only thing I think is left is this unclear line in multi-instance.rst:

> A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instances) is generated when **case.setup** is called.

Can we please rephrase it to:

A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instance) is generated for each component instance when **case.setup** is called.

mvertens commented:
I agree that an in-person chat would be useful. I'd like to be there as
well on Tuesday.

On Sun, Sep 3, 2017 at 6:54 AM, Bill Sacks <notifications@github.com> wrote:

> *@billsacks* commented on this pull request.
> ------------------------------
>
> In doc/source/users_guide/multi-instance.rst
> <https://github.com/ESMCI/cime/pull/1821#discussion_r136715991>:
>
> >
> -#. Multiple instances generally should un concurrently, which is the default setting in **env_mach_pes.xml**.
> -   The serial setting is only for EXPERT USERS in upcoming development code implementations.
> +#. In multi-driver mode you will always get 1 instance of each component for each driver/coupler, if you change a case using xmlchange MULTI_COUPLER=TRUE you will get a number of driver/couplers equal to the maximum NINST value over all components.
>
> Sorry, I took some time to try to understand the current behavior and
> expected behavior, but I'm still a bit lost. Could we have a brief
> in-person chat about this on Tuesday?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMCI/cime/pull/1821#discussion_r136715991>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHlxE17qup0zwbaF8Ssko3patP9w9UgSks5seqFogaJpZM4O5i5g>
> .
>

mvertens commented:
@gold2718, @jedwards4b -  I think this is a really good idea and one that
should be implemented to provide a robust implementation of this feature.

On Sun, Sep 3, 2017 at 1:16 PM, goldy <notifications@github.com> wrote:

> *@gold2718* commented on this pull request.
> ------------------------------
>
> In doc/source/users_guide/multi-instance.rst
> <https://github.com/ESMCI/cime/pull/1821#discussion_r136723998>:
>
> >
> -#. Multiple instances generally should un concurrently, which is the default setting in **env_mach_pes.xml**.
> -   The serial setting is only for EXPERT USERS in upcoming development code implementations.
> +#. In multi-driver mode you will always get 1 instance of each component for each driver/coupler, if you change a case using xmlchange MULTI_COUPLER=TRUE you will get a number of driver/couplers equal to the maximum NINST value over all components.
>
> If we decide that there should be a fix, I propose the following actions:
>
>    1. When setting MULTI_DRIVER from FALSE to TRUE, check to make sure
>    all NINST_xxx have the same value and generate an error if not.
>    2. If MULTI_DRIVER == TRUE, then case.setup will check that all
>    NINST_xxx have the same value and generate an error if not.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMCI/cime/pull/1821#discussion_r136723998>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHlxEyy0vVt6pHsmTtsPv0qytD1HpKavks5sevsSgaJpZM4O5i5g>
> .
>

jedwards4b commented:
@billsacks @gold2718 I think that we need to sit down and talk about this when we can - it's not clear that we can put in the kind of error checking that you are requesting without creating new problems.  Suppose for example I add an error check that NINST_XXX has to be the same for all components in MULTI_DRIVER mode.   Now someone is in that mode and wants to change the NINST value from 3 to 4.   Should ```xmlchange NINST_ATM=4``` be an error (since all the others are still 3)?

gold2718 commented:
@jedwards4b,
I'm sick and not sure when I will be in next.
However, your example is why I did _not_ call for an NINST check in xmlchange. If you look at my two suggestions, they catch inconsistencies without (I think) leading to unwanted outcomes. Please review them (copied below for your convenience):

1. When setting MULTI_DRIVER from FALSE to TRUE, check to make sure all NINST_xxx have the same value and generate an error if not.
2. If MULTI_DRIVER == TRUE, then case.setup will check that all NINST_xxx have the same value and generate an error if not.

jedwards4b commented:
1. Is easy, but I'm not sure it should be done because
2. leads to a situation wherein if I have MULTI_DRIVER=TRUE I cannot change the NINST_XXX values without generating an error.     We can talk when you are feeling better.

gold2718 commented:
I do not see how 2 leads to the situation. The only new xmlchange check is on setting MULTI_DRIVER from FALSE to TRUE. There is no consistency check when setting NINST_xxx.

mvertens commented:
@gold2718 @jedwards @billsacks - at this point I think a quick face to face
meeting once @gold2718 feels better and is back will be the most productive
way to resolve this issue. I think this can be a short meeting where we
will be able reach consensus quickly on the best way forwards.

On Tue, Sep 5, 2017 at 1:52 PM, goldy <notifications@github.com> wrote:

> I do not see how 2 leads to the situation. The only new xmlchange check is
> on setting MULTI_DRIVER from FALSE to TRUE. There is no consistency check
> when setting NINST_xxx.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMCI/cime/pull/1821#issuecomment-327284439>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHlxE3e_8wfCMmq_xZUvND8yveC8SoXPks5sfaaAgaJpZM4O5i5g>
> .
>

jedwards4b commented:
@gold2718 I think I've figured out how to add the error check you requested.   Have another look.

gold2718 commented:
Feeling pretty good about this now.

jedwards4b commented:
I've rerun scripts_regression_tests.py and fixed a pylint issue.  @billsacks are you still reviewing?

billsacks commented:
I don't feel a need to review this.
